googleapiclient/discovery_cache/file_cache.py

```diff
@@ -0,0 +1,114 @@
+# Copyright 2014 Google Inc. All Rights Reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#      http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+"""File based cache for the discovery document."""
+
+import datetime
+import json
+import logging
+import os
+import tempfile
+import threading
+
+from oauth2client.locked_file import LockedFile
+
+from . import base
+from ..discovery_cache import DISCOVERY_DOC_MAX_AGE
+
+logger = logging.getLogger(__name__)
+
+FILENAME = 'discovery-doc.cache'
+EPOCH = datetime.datetime.utcfromtimestamp(0)
+
+
+def to_timestamp(d):
+  return (d - EPOCH).total_seconds()
+
+
+class Cache(base.Cache):
+  """A file based cache for the discovery documents.
+
+  Attributes:
+    max_age: Cache expiration in seconds.
+  """
+
+  def __init__(self, max_age):
+      """Constructor for appengine_memcache.Cache.
+
+      Args:
+        max_age: Cache expiration in seconds.
+      """
+      self.max_age = max_age
+      self._file = os.path.join(tempfile.gettempdir(), FILENAME)
+      f = LockedFile(self._file, 'a+', 'r')
+      try:
+        f.open_and_lock()
+        if f.is_locked():
+          try:
+            f.file_handle().seek(0)
+            d = json.load(f.file_handle())
+          except Exception as e:
+            # This means we opened this file for the first time, or
+            # there is a file corruption. Either case, we can
+            # initialize it with an empty dictionary.
+            f.file_handle().truncate(0)
+            f.file_handle().seek(0)
+            json.dump({}, f.file_handle())
+        # If we can not obtain the lock, other process or thread must
+        # have initialized the file.
+      except Exception as e:
+        logging.warning(e, exc_info=True)
+      finally:
+        f.unlock_and_close()
+
+  def get(self, url):
+    f = LockedFile(self._file, 'r+', 'r')
+    try:
+      f.open_and_lock()
+      if f.is_locked():
+        f.file_handle().seek(0)
+        cache = json.load(f.file_handle())
+        if url in cache:
+          content, t = cache.get(url, (None, 0))
+          if to_timestamp(datetime.datetime.now()) < t + self.max_age:
+            return content
+        return None
+      else:
+        logger.debug('Could not obtain a lock for the cache file.')
+        return None
+    except Exception as e:
+      logger.warning(e, exc_info=True)
+    finally:
+      f.unlock_and_close()
+
+  def set(self, url, content):
+    f = LockedFile(self._file, 'r+', 'r')
```

Comment:
I'm confused by the lack of "w" on this line - doesn't this method have to write to the file?